File: CSharp/Calculator/Token.cs
/*--------------------------------------------------------------------------*\
::
::  Copyright © 2023 Steffen Liersch
::  https://www.steffen-liersch.de/
::
\*--------------------------------------------------------------------------*/

using System;
using System.Globalization;

namespace Liersch.Calculator;

sealed class Token
{
  public readonly double? AsNumber;

  public readonly string AsString;

  public Token(double value)
  {
    AsNumber = value;
    AsString = value.ToString(CultureInfo.InvariantCulture);
  }

  public Token(string value)
  {
    if(double.TryParse(value, NumberStyles.Float, CultureInfo.InvariantCulture, out double v))
    {
      AsNumber = v;
      AsString = value;
    }
    else
    {
      switch(value.ToUpperInvariant())
      {
        case "E":
          AsNumber = Math.E;
          AsString = "E";
          break;

        case "PI":
          AsNumber = Math.PI;
          AsString = "PI";
          break;

        default:
          AsString = value;
          break;
      }
    }
  }

  public override string ToString() => AsString;
}

File: Java/src/main/java/de/steffen_liersch/calculator/Token.java
/*--------------------------------------------------------------------------*\
::
::  Copyright © 2023 Steffen Liersch
::  https://www.steffen-liersch.de/
::
\*--------------------------------------------------------------------------*/

package de.steffen_liersch.calculator;

final class Token
{
  public final Double asNumber;

  public final String asString;

  public Token(double value)
  {
    asNumber = value;
    asString = String.valueOf(value);
  }

  public Token(String value)
  {
    Double d = tryParseDouble(value);
    if(d != null)
    {
      asNumber = d;
      asString = value;
    }
    else
    {
      switch(value.toUpperCase())
      {
        case "E":
          asNumber = Math.E;
          asString = "E";
          break;

        case "PI":
          asNumber = Math.PI;
          asString = "PI";
          break;

        default:
          asNumber = null;
          asString = value;
          break;
      }
    }
  }

  @Override
  public String toString()
  {
    if(asNumber != null && asNumber < 0)
      return "(" + asNumber + ")";
    return asString;
  }

  private static Double tryParseDouble(String value)
  {
    try
    {
      return Double.parseDouble(value);
    }
    catch(NumberFormatException e)
    {
      return null;
    }
  }
}
